go/glome/glome.go

```diff
@@ -0,0 +1,199 @@
+// from the user to the peer. Sets MinPeerTagSize as MaxTagSize.
+func (user *PrivateKey) Exchange(peer *PublicKey) (*Dialog, error) {
+	slice, err := curve25519.X25519(user[:], peer[:])
+	if err != nil {
+		return nil, err
+	}
+
+	public, err := user.Public()
+	if err != nil {
+		return nil, err
+	}
+
+	var s SharedSecret
+	copy(s[:], slice)
+	return &Dialog{shared: s, User: *public, Peer: *peer, MinPeerTagSize: MaxTagSize}, nil
+}
+
+// Generates a Dialog struct. It performs GLOME handshake, and stores create a Dialog
+// from the user to the peer. Sets param m as MinPeerTagSize.
+func (user *PrivateKey) TruncatedExchange(peer *PublicKey, m uint) (*Dialog, error) {
+
+	if m == 0 || m > MaxTagSize {
+		return nil, ErrInvalidTagSize
+	}
+
+	d, err := user.Exchange(peer)
+	if err != nil {
+		return nil, err
+	}
+
+	d.MinPeerTagSize = m
+	return d, nil
+}
+
+// Generates a private key from a slice. Fail if len of slice is not PrivateKeySize
+func PrivateKeyFromSlice(b []byte) (*PrivateKey, error) {
+	if len(b) != PrivateKeySize {
+		return nil, ErrInvalidPrivateKey
+	}
+
+	var p PrivateKey
+	copy(p[:], b)
+	return &p, nil
+}
+
+// GenerateKey generates a public/private key pair using entropy from rand.
+func GenerateKeys(rand io.Reader) (*PublicKey, *PrivateKey, error) {
+	b := make([]byte, PrivateKeySize)
+	_, err := rand.Read(b)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	priv, err := PrivateKeyFromSlice(b)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pub, err := priv.Public()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return pub, priv, nil
+}
+
+// Implement tag managing functionalities for GLOME protocol.
+type Dialog struct {
+	shared         SharedSecret
+	User           PublicKey // User's Public key
+	Peer           PublicKey // Peer's Public key
+	MinPeerTagSize uint      // Minimun Tag size allowed.
+}
+
+func (d *Dialog) sendingKey() []byte {
+	return append(d.shared[:], append(d.Peer[:], d.User[:]...)...)
+}
+
+func (d *Dialog) receivingKey() []byte {
+	return append(d.shared[:], append(d.User[:], d.Peer[:]...)...)
+}
+
+// Generates a tag matching some provided message, counter and password.
+func tag(msg []byte, c uint8, password []byte) []byte {
+	h := hmac.New(sha256.New, password)
+	h.Write([]byte{c})
+	h.Write(msg)
+	return h.Sum(nil)
+}
+
+// Generates a tag matching some provided message and counter.
+// This tag is generated following GLOME protocol specification
+// in the context of a communication from the users to theirs peers.
+func (d *Dialog) Tag(msg []byte, c uint8) []byte {
+	return tag(msg, c, d.sendingKey())
+}
+
+// Checks if a tag matches some provided message and counter.
+// The method generates the matching tag following GLOME protocol
+// specification in the context of a communication from the users'
+// peers to the users and then is compared with the tag provided.
+func (d *Dialog) Check(t []byte, msg []byte, c uint8) bool {
```

Comment:
Method parameters in Go should not be short, as they are part of the documentation. I'd suggest to s/t/tag/ and s/c/counter/ everywhere. s/msg/message/ is debatable but might make sense for clarity.

Reply:
Fixed.